### PR TITLE
add ganglia emitter extension

### PR DESCRIPTION
**Pull request by hellobabygogo**
add ganglia emitter extension

### CODE REVIEW

**Comment by dkhwangbo** [comment]
@hellobabygogo 
Hi. it would be great if you can write some documentation for this extension. For example, here's the documentation for Kafka emitter
 - https://github.com/apache/incubator-druid/blob/master/docs/content/development/extensions-contrib/kafka-emitter.md

**Comment by hellobabygogo** [comment]
> @hellobabygogo
> Hi. it would be great if you can write some documentation for this extension. For example, here's the documentation for Kafka emitter
> 
> * https://github.com/apache/incubator-druid/blob/master/docs/content/development/extensions-contrib/kafka-emitter.md

ok

File: docs/content/development/extensions-contrib/ganglia-emitter.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+|--------|-----------|---------|-------|
+|`druid.emitter.ganglia.hostname`|The hostname of the Ganglia server.|yes|none|
+|`druid.emitter.ganglia.port`|The port of the Ganglia server.|yes|none|
+|`druid.emitter.ganglia.separator`|Metric name separator|no|.|
+|`druid.emitter.ganglia.includeHost`|Flag to include the hostname as part of the metric name.|no|false|
+|`druid.emitter.ganglia.dimensionMapPath`|JSON file defining the Ganglia type, and desired dimensions for every Druid metric|no|Default mapping provided. See below.|
+|`druid.emitter.ganglia.blankHolder`|The blank character replacement as Ganglia does not support path with blank character|no|"-"|
+|`druid.emitter.ganglia.flushPeriod` | Queue flushing period in milliseconds. |no|1 minute|
+|`druid.emitter.ganglia.loadPeriod` | Load metrics config period in milliseconds. |no|1 minute|
+|`druid.emitter.ganglia.maxQueueSize`| Maximum size of the queue used to buffer events. |no|`MAX_INT`|
```
Comment (dclim) [inline docs/content/development/extensions-contrib/ganglia-emitter.md]: `Integer.MAX_VALUE` isn't a suitable default for the queue size. This means that if there are issues emitting to the Ganglia server, the buffer will continue to fill up until it consumes the entire heap and crashes the server.

The emitter should aim to minimize the performance impact it has on the Druid process. Attempting to limit memory usage to something like 10% of the heap would be more appropriate.

File: docs/content/development/extensions-contrib/ganglia-emitter.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+|`druid.emitter.ganglia.emitWaitTime` | wait time in milliseconds to try to send the event otherwise emitter will throwing event. |no|0|
```
Comment (dclim) [inline docs/content/development/extensions-contrib/ganglia-emitter.md]: This doesn't seem to be the wait time for sending the event, but rather the wait time for the blocking queue to accept the event. Maybe reword this as something like 'Wait time in milliseconds for send buffer to accept before dropping an event'

File: extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+          if (gangliaEvent != null) {
+            try {
+              ganglia.announce(
+                  gangliaEvent.getName(),
+                  gangliaEvent.getValue().longValue(),
+                  gangliaEvent.getService()
+              );
+            }
+            catch (GangliaException e) {
+              log.error(e, "Dropping event [%s]. Unable to send to Ganglia.", gangliaEvent);
```
Comment (dclim) [inline extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java]: Would this generate an excessive number of messages (with ERROR severity) whenever there's an issue, maybe for example a temporary network glitch preventing communication with Ganglia? Consider logging the first one (at info level) and then suppressing logging on the rest of them.

File: extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/DimensionConverter.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import com.google.common.collect.ImmutableList;
+import org.apache.druid.java.util.common.logger.Logger;
+
+import java.util.Map;
+
+public class DimensionConverter
+{
+  private static final Logger log = new Logger(DimensionConverter.class);
+
+  public GangliaMetric addFilteredUserDims(
```
Comment (dclim) [inline extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/DimensionConverter.java]: This can be static

File: extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.regex.Pattern;
+
+/**
+ * Created by yangxuan on 2018/9/6.
```
Comment (dclim) [inline extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java]: Can you remove this comment? We don't include author headers

File: extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+ */
+public class GangliaEmitter implements Emitter
+{
+  private static final Logger log = new Logger(GangliaEmitter.class);
+  private static final char DRUID_METRIC_SEPARATOR = '/';
+  private static final Pattern STATSD_SEPARATOR = Pattern.compile("[:|]");
+  private static final Pattern BLANK = Pattern.compile("\\s+");
+
+  private final GangliaEmitterConfig config;
+  private final MetricRegistry registry;
```
Comment (dclim) [inline extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java]: This is never used?

File: extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+  private GMetric ganglia;
+  private final DimensionConverter converter;
+  private volatile ScheduledExecutorService scheduledExecutor;
+  private final ObjectMapper mapper;
+  private final LinkedBlockingQueue<GangliaEvent> eventsQueue;
+  private AtomicLong countLostEvents = new AtomicLong(0);
+  private final ScheduledExecutorService exec = Executors.newScheduledThreadPool(2, new ThreadFactoryBuilder()
```
Comment (dclim) [inline extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java]: Why 2? A single thread should be fine for this scheduler

File: extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+  private static final Logger log = new Logger(GangliaEmitter.class);
+  private static final char DRUID_METRIC_SEPARATOR = '/';
+  private static final Pattern STATSD_SEPARATOR = Pattern.compile("[:|]");
+  private static final Pattern BLANK = Pattern.compile("\\s+");
+
+  private final GangliaEmitterConfig config;
+  private final MetricRegistry registry;
+  private GMetric ganglia;
+  private final DimensionConverter converter;
+  private volatile ScheduledExecutorService scheduledExecutor;
```
Comment (dclim) [inline extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java]: Do you need a second scheduled exec here or can you re-use `exec`? Does this need to be volatile?

File: extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+          new Runnable()
+          {
+            @Override
+            public void run()
+            {
+              try {
+                metricMap = readMap(mapper, config.getDimensionMapPath());
+              }
+              catch (Exception e) {
+                log.error(e.toString());
```
Comment (dclim) [inline extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java]: Maybe better to log the entire stack trace?

File: extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+      //load config
+      scheduledExecutor = Execs.scheduledSingleThreaded("ganglia_metrics_load__scheduled_%d");
+      scheduledExecutor.scheduleAtFixedRate(
+          new Runnable()
+          {
+            @Override
+            public void run()
+            {
+              try {
+                metricMap = readMap(mapper, config.getDimensionMapPath());
```
Comment (dclim) [inline extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java]: You don't strictly need to pass the `mapper` in here, it's already visible to the method from the instance field

File: extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitterConfig.java [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+
+package org.apache.druid.emitter.ganglia;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+
+public class GangliaEmitterConfig
+{
+  private static final Long DEFAULT_LOAD_PERIOD = (long) (60 * 1000); // load every one minute
```
Comment (dclim) [inline extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitterConfig.java]: Instead of casting, do `60L * 1000` or similar

File: extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+      String metric = metricEvent.getMetric();
+      Map<String, Object> userDims = metricEvent.getUserDims();
+      ImmutableList.Builder<String> nameBuilder = new ImmutableList.Builder<>();
+      if (config.getIncludeHost()) {
+        nameBuilder.add(host);
+      }
+      nameBuilder.add(service);
+      GangliaMetric gangliaMetric = converter.addFilteredUserDims(service, metric, userDims, nameBuilder, metricMap);
+      if (gangliaMetric == null) {
+        log.debug("Metric=[%s] has no ganglia type mapping", gangliaMetric);
```
Comment (dclim) [inline extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java]: `gangliaMetric` will always be null here

File: extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+        return;
+      }
+      nameBuilder.add(metric);
+      String fullName = Joiner.on(config.getSeparator())
+                              .join(nameBuilder.build());
+      fullName = StringUtils.replaceChar(fullName, DRUID_METRIC_SEPARATOR, config.getSeparator());
+      fullName = STATSD_SEPARATOR.matcher(fullName).replaceAll(config.getSeparator());
+      fullName = BLANK.matcher(fullName).replaceAll(config.getBlankHolder());
+      GangliaEvent gangliaEvent = new GangliaEvent(fullName, metricEvent.getValue(), StringUtils.replaceChar(service, DRUID_METRIC_SEPARATOR, config.getSeparator()));
```
Comment (dclim) [inline extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java]: Can you format this so it doesn't exceed 120 characters on one line?

File: extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+      try {
+        final boolean successful = eventsQueue.offer(
+            gangliaEvent,
+            config.getEmitWaitTime(),
+            TimeUnit.MILLISECONDS
+        );
+        if (!successful) {
```
Comment (dclim) [inline extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java]: Nested ifs can be put into a single statement

File: extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+          if (countLostEvents.getAndIncrement() % 1000 == 0) {
+            log.error(
+                "Lost total of [%s] events because of emitter queue is full. Please increase the capacity or/and the consumer frequency",
+                countLostEvents.get()
+            );
+          }
+        }
+      }
+      catch (InterruptedException e) {
+        log.error(e, "Emitter metric interrupted with message [%s]", e.getMessage());
```
Comment (dclim) [inline extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java]: If you're logging the stack trace, you don't need to log the message again

File: extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  @Override
+  public void flush() throws IOException
```
Comment (dclim) [inline extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java]: Can `flush()` force an immediate send to Ganglia?

File: extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+    ganglia.close();
+  }
+
+  private Map<String, GangliaMetric> readMap(ObjectMapper mapper, String dimensionMapPath)
+  {
+    try {
+      InputStream is;
+      if (Strings.isNullOrEmpty(dimensionMapPath)) {
+        log.info("Using default ganglia metric dimension and types");
+        is = this.getClass().getClassLoader().getResourceAsStream("defaultMetricDimensions.json");
```
Comment (dclim) [inline extensions-contrib/ganglia-emitter/src/main/java/org/apache/druid/emitter/ganglia/GangliaEmitter.java]: Can you name this file with Ganglia and emitter in the name?

File: docs/content/development/extensions-contrib/ganglia-emitter.md (hunk 1 above)
Comment (dclim) [reply docs/content/development/extensions-contrib/ganglia-emitter.md]: Also, the buffer should be size based rather than number of items based for more predictable memory usage. Take a look at `MemoryBoundLinkedBlockingQueue` used by KafkaEmitter and if you can use it, it may make sense to move it to core Druid and reuse it.

**Comment by dclim** [comment]
@hellobabygogo have you had a chance to review the PR comments?

**Comment by hellobabygogo** [comment]
> @hellobabygogo have you had a chance to review the PR comments?

@dclim hello, Thanks for your suggestions, I will revise it soon.
